_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+
+Implementing a microservices architecture in Java isn't for everyone. For that matter, implementing microservices, in general, isn't often needed. Most companies do it to scale their people, not their systems. Even Martin Fowler's original blog post on [Microservices](https://martinfowler.com/articles/microservices.html) recommends against it:
+
+> One reasonable argument we've heard is that you shouldn't start with a microservices architecture. Instead begin with a monolith, keep it modular, and split it into microservices once the monolith becomes a problem.
+
+The Java ecosystem has some well-established patterns for developing microservice architectures. If you're familiar with Spring, you'll feel right at home developing with Spring Boot and Spring Cloud. Since that's one of the quickest ways to get started, I figured I'd walk you through a quick example.
+
+This example contains a microservice with a REST API that returns a list of cool cars. It uses Netflix Eureka for service discovery, WebClient for remote communication, and Spring Cloud Gateway to route requests to the microservice. It integrates Spring Security and OAuth 2.0, so only authenticated users can access the API gateway and microservice. It also uses Resilience4j to add fault tolerance to the gateway.
+
+{% img blog/spring-boot-microservices/spring-cloud-gateway-with-cars.png alt:"Spring Boot Microservices" %}{: .center-image }
+
+**Table of Contents**{: .hide }
+* Table of Contents
+{:toc}
+
+## Create Java Microservices with Spring Boot and Spring Cloud
+
+I like to show developers how to build everything from scratch. Today, I'm going to take a different approach. First, I'll show you how to get the completed example working. Then, I'll explain how I created everything and the trials and tribulations I encountered along the way.
+
+You can start by cloning the [@oktadev/auth0-java-microservices-examples](https://github.com/oktadev/auth0-java-microservices-examples) repository.
+
+```shell
+git clone https://github.com/oktadev/auth0-java-microservices-examples.git
+cd auth0-java-microservices-examples/spring-boot-gateway-webflux
+```
+
+In the `spring-boot-gateway-webflux` directory, there are three projects:
+
+* **discovery-service**: a Netflix Eureka server used for service discovery.
+* **car-service**: a simple Car Service that uses Spring Data REST to serve up a REST API of cars.
+* **api-gateway**: an API gateway with a `/cool-cars` endpoint that talks to the car service and filters out cars that aren't cool (in my opinion, of course).
+
+### Run a Secure Spring Boot Microservice Architecture
+
+To run the example, you must [install the Auth0 CLI](https://github.com/auth0/auth0-cli#installation) and create an Auth0 account. If you don't have an Auth0 account, [sign up for free](https://auth0.com/signup). I recommend using [SDKMAN!](https://sdkman.io) to install Java 17+.
```

Comment:
would be nice to mention HTTPie also here since its used later and someone could be confused